lib/std/mem.zig

```diff
@@ -630,6 +631,190 @@ test "spanZ" {
     try testing.expectEqual(@as(?[:0]u16, null), spanZ(@as(?[*:0]u16, null)));
 }
 
+/// Helper for the return type of sliceTo()
+fn SliceTo(comptime T: type, comptime end: meta.Elem(T)) type {
+    switch (@typeInfo(T)) {
+        .Optional => |optional_info| {
+            return ?SliceTo(optional_info.child, end);
+        },
+        .Pointer => |ptr_info| {
+            var new_ptr_info = ptr_info;
+            new_ptr_info.size = .Slice;
+            switch (ptr_info.size) {
+                .One => switch (@typeInfo(ptr_info.child)) {
+                    .Array => |array_info| {
+                        new_ptr_info.child = array_info.child;
+                        // The return type must only be sentinel terminated if we are guaranteed
+                        // to find the value searched for, which is only the case if it matches
+                        // the sentinel of the type passed.
+                        if (array_info.sentinel) |sentinel| {
+                            if (end == sentinel) {
+                                new_ptr_info.sentinel = end;
+                            } else {
+                                new_ptr_info.sentinel = null;
+                            }
+                        }
+                    },
+                    else => {},
```

Comment:
should this be an compile error/unreachable?

Reply:
It is, see the `@compileError()` at the bottom of the function.

Reply:
no it doesn't? won't this hit the `return @Type(std.builtin.TypeInfo{ .Pointer = new_ptr_info });`?

Reply:
Ah yes you're right but this is technically fine as `sliceToLen()` will give a compile error. We rely on this for non sentinel terminated "many" pointers as well currently.